### PR TITLE
manifest create,add,inspect

### PR DESCRIPTION
**Pull request by QiWang19**
Implememts manifest subcommands create, add, inspect.

Signed-off-by: Qi Wang <qiwan@redhat.com>

### CODE REVIEW

File: cmd/podmanV2/manifest/add.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+func add(cmd *cobra.Command, args []string) error {
+	var (
+		listImageSpec string
+		imageSpec     string
+	)
+	switch len(args) {
+	case 0, 1:
+		return errors.New("At least a list image and an image to add must be specified")
```
Comment (jwhonce) [inline cmd/podmanV2/manifest/add.go:47]:
```suggestion
```
Args: cobra.MinimumNArgs(2) Ensures this case will never be true.

File: cmd/podmanV2/manifest/add.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+		listImageSpec = args[0]
+		if listImageSpec == "" {
+			return errors.Errorf(`Invalid image name "%s"`, args[0])
+		}
+		imageSpec = args[1]
+		if imageSpec == "" {
+			return errors.Errorf(`Invalid image name "%s"`, args[1])
+		}
+	default:
+		return errors.New("At least two arguments are necessary: list and image to add to list")
```
Comment (jwhonce) [inline cmd/podmanV2/manifest/add.go:58]: Given this default, I would change to Args.ExactArgs(2) and remove the switch.

File: cmd/podmanV2/manifest/create.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+		Parent:  manifestCmd,
+	})
+	flags := createCmd.Flags()
+	flags.BoolVar(&manifestCreateOpts.All, "all", false, "add all of the lists' images if the images to add are lists")
+}
+
+func create(cmd *cobra.Command, args []string) error {
+	if len(args) == 0 {
+		return errors.New("At least a name must be specified for the list")
+	}
```
Comment (jwhonce) [inline cmd/podmanV2/manifest/create.go:39]:
```suggestion
``
Args: cobra.MinimumNArgs(1) ensures this is always false.

File: cmd/podmanV2/manifest/inspect.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+	switch len(args) {
+	case 0:
+		return errors.New("At least a source list ID must be specified")
+	case 1:
+		if args[0] == "" {
+			return errors.Errorf(`Invalid image name "%s"`, args[0])
+		}
+	default:
+		return errors.New("Only one argument is necessary for inspect: an image name")
+	}
```
Comment (jwhonce) [inline cmd/podmanV2/manifest/inspect.go:40]:
```suggestion
```
Args: cobra.ExactArgs(1) will take care of this switch

File: cmd/podmanV2/manifest/manifest.go [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+)
+
+var (
+	manifestDescription = "Creates, modifies, and pushes manifest lists and image indexes."
+	manifestCmd         = &cobra.Command{
+		Use:               "manifest",
+		Short:             "Manipulate manifest lists and image indexes",
+		Long:              manifestDescription,
+		TraverseChildren:  true,
+		PersistentPreRunE: preRunE,
```
Comment (jwhonce) [inline cmd/podmanV2/manifest/manifest.go:16]:
```suggestion
```
As of https://github.com/containers/libpod/pull/5848 this is no longer required.

File: cmd/podmanV2/manifest/manifest.go [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+		RunE:              registry.SubCommandExists,
+		Example: `podman manifest create localhost/list
+  podman manifest inspect localhost/list`,
+	}
+)
+
+func init() {
+
```
Comment (jwhonce) [inline cmd/podmanV2/manifest/manifest.go:24]:
```suggestion
```

File: cmd/podmanV2/manifest/manifest.go [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+	registry.Commands = append(registry.Commands, registry.CliCommand{
+		Mode:    []entities.EngineMode{entities.ABIMode, entities.TunnelMode},
+		Command: manifestCmd,
+	})
+	manifestCmd.SetHelpTemplate(registry.HelpTemplate())
+	manifestCmd.SetUsageTemplate(registry.UsageTemplate())
```
Comment (jwhonce) [inline cmd/podmanV2/manifest/manifest.go:30]:
```suggestion
```

File: cmd/podmanV2/manifest/manifest.go [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+// preRunE populates the image engine for sub commands
+func preRunE(cmd *cobra.Command, args []string) error {
+	_, err := registry.NewImageEngine(cmd, args)
+	return err
+}
```
Comment (jwhonce) [inline cmd/podmanV2/manifest/manifest.go:37]:
```suggestion
```
No longer required.

File: pkg/domain/infra/abi/manifest.go [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+	"github.com/pkg/errors"
+)
+
+// ManifestCreate implements logic for creating manifest lists via ImageEngine
+func (ir *ImageEngine) ManifestCreate(ctx context.Context, names, images []string, opts entities.ManifestCreateOptions) error {
+	fullNames, err := buildahUtil.ExpandNames(names, "", ir.Libpod.SystemContext(), ir.Libpod.GetStore())
+	if err != nil {
+		return errors.Wrapf(err, "error encountered while expanding image name %q", names)
+
```
Comment (jwhonce) [inline pkg/domain/infra/abi/manifest.go:25]:
```suggestion
```

File: pkg/domain/infra/abi/manifest.go [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+	}
+	imageID, err := libpodImage.CreateManifestList(ir.Libpod.ImageRuntime(), *ir.Libpod.SystemContext(), fullNames, images, opts.All)
+	if err == nil {
+		fmt.Printf("%s\n", imageID)
+	}
```
Comment (jwhonce) [inline pkg/domain/infra/abi/manifest.go:30]: Cannot use Printf() in this layer.  Need to return data to caller.

File: pkg/domain/infra/abi/manifest.go [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+	return err
+}
+
+// ManifestInspect returns the content of a manifest list or image
+func (ir *ImageEngine) ManifestInspect(ctx context.Context, name string) error {
+	dockerPrefix := fmt.Sprintf("%s://", docker.Transport.Name())
+	_, err := alltransports.ParseImageName(name)
+	if err != nil {
+		_, err = alltransports.ParseImageName(fmt.Sprintf("%s%s", dockerPrefix, name))
```
Comment (jwhonce) [inline pkg/domain/infra/abi/manifest.go:39]:
```suggestion
		_, err = alltransports.ParseImageName(dockerPrefix + name)
```

File: pkg/domain/infra/abi/manifest.go [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+
+	list, err := image.InspectManifest()
+	if err != nil {
+		return errors.Wrapf(err, "loading manifest %q", name)
+	}
+	buf, err := json.MarshalIndent(list, "", "    ")
+	if err != nil {
+		return errors.Wrapf(err, "error rendering manifest for display")
+	}
+	fmt.Printf("%s\n", buf)
```
Comment (jwhonce) [inline pkg/domain/infra/abi/manifest.go:57]:
```suggestion
return buf, nil
```
This needs to be returned to caller, not printed at this layer.

File: pkg/domain/entities/engine_image.go [hunk 13]
```diff
@@ -23,4 +23,7 @@ type ImageEngine interface {
 	Tag(ctx context.Context, nameOrId string, tags []string, options ImageTagOptions) error
 	Untag(ctx context.Context, nameOrId string, tags []string, options ImageUntagOptions) error
 	Search(ctx context.Context, term string, opts ImageSearchOptions) ([]ImageSearchReport, error)
+	ManifestCreate(ctx context.Context, names, images []string, opts ManifestCreateOptions) error
+	ManifestInspect(ctx context.Context, name string) error
+	ManifestAdd(ctx context.Context, opts ManifestAddOptions) error
```
Comment (jwhonce) [inline pkg/domain/entities/engine_image.go:28]:
```suggestion
	ManifestCreate(ctx context.Context, names, images []string, opts ManifestCreateOptions) (<report>, error)
	ManifestInspect(ctx context.Context, name string)  (<report>, error)
	ManifestAdd(ctx context.Context, opts ManifestAddOptions)  (<report>, error)
```
You cannot present data to user in these functions, needs to be done in the manifest/*.go files.

File: test/e2e/manifest_test.go [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+		imageListAMD64InstanceDigest   = "sha256:59eec8837a4d942cc19a52b8c09ea75121acc38114a2c68b98983ce9356b8610"
+		imageListARMInstanceDigest     = "sha256:c84b0a3a07b628bc4d62e5047d0f8dff80f7c00979e1e28a821a033ecda8fe53"
+		imageListPPC64LEInstanceDigest = "sha256:bcf9771c0b505e68c65440474179592ffdfa98790eb54ffbf129969c5e429990"
+		imageListS390XInstanceDigest   = "sha256:882a20ee0df7399a445285361d38b711c299ca093af978217112c73803546d5e"
+	)
+
+	BeforeEach(func() {
+		tempdir, err = CreateTempDirInTempDir()
+		if err != nil {
+			os.Exit(1)
```
Comment (jwhonce) [inline test/e2e/manifest_test.go:31]: You should print out err to help with debugging issues.

**Comment by QiWang19** [comment]
@jwhonce PTAL

File: test/e2e/manifest_test.go (hunk 14 above)
Comment (QiWang19) [reply test/e2e/manifest_test.go:33]: I checked other tests. Integration tests don't log errors in BeforeEach(). Does this need for manifest?

**Comment by QiWang19** [comment]
@jwhonce @rhatdan @mheon @TomSweeneyRedHat PTAL

File: completions/bash/podman [hunk 15]
```diff
@@ -1733,6 +1733,83 @@ _podman_logs() {
...
+     esac
+}
+
+_podman_manifest_add() {
+	local options_with_args="
+      --annotation
+      --arch
+	  --features
+	  --os-version
+	  --variant
```
Comment (TomSweeneyRedHat) [inline completions/bash/podman:1764]: Did you meant to have an extra couple of spaces before the last three lines here?

File: completions/bash/podman [hunk 16]
```diff
@@ -1733,6 +1733,83 @@ _podman_logs() {
...
+        *)
+            __podman_complete_images --id
+            ;;
+    esac
+}
+
+_podman_manifest_create() {
+  local boolean_options="
+    --all
+	--help
```
Comment (TomSweeneyRedHat) [inline completions/bash/podman:1787]: spacing?

File: completions/bash/podman [hunk 17]
```diff
@@ -3356,6 +3433,7 @@ _podman_podman() {
     login
     logout
     logs
+	manifest
```
Comment (TomSweeneyRedHat) [inline completions/bash/podman:3436]: spacing?

**Comment by TomSweeneyRedHat** [comment]
A couple of spacing/tab nits, otherwise LGTM

File: cmd/podman/manifest/add.go [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+func init() {
+	registry.Commands = append(registry.Commands, registry.CliCommand{
+		Mode:    []entities.EngineMode{entities.ABIMode, entities.TunnelMode},
+		Command: addCmd,
+		Parent:  manifestCmd,
+	})
+	flags := addCmd.Flags()
+	flags.BoolVar(&manifestAddOpts.All, "all", false, "add all of the list's images if the image is a list")
+	flags.StringSliceVar(&manifestAddOpts.Annotation, "annotation", nil, "set an `annotation` for the specified image")
+	flags.StringVar(&manifestAddOpts.Arch, "arch", "", "override the `architecture` of the specified image")
```
Comment (rhatdan) [inline cmd/podman/manifest/add.go:35]: Buildah has --os as well?

**Comment by rhatdan** [comment]
Buildah also has
buildah-manifest-annotate.1  buildah-manifest-push.1
buildah-manifest-remove.1

File: cmd/podman/manifest/add.go (hunk 18 above)
Comment (QiWang19) [reply cmd/podman/manifest/add.go:35]: Need to change libpod/image/manifests.go to support `--os`, I can fix it in another PR.

**Comment by QiWang19** [comment]
> Buildah also has
> buildah-manifest-annotate.1 buildah-manifest-push.1
> buildah-manifest-remove.1

I plan to finish the rest implementation in a different PR.

**Pull request merged**